crates/context_aware_config/src/api/custom_types/mod.rs

```diff
@@ -0,0 +1,104 @@
+mod types;
+
+use crate::db::models::JsonSchemaTypes;
+use crate::db::schema::jsonschema_types::{self, dsl};
+use actix_web::web::{Json, Path};
+use actix_web::{delete, get, put, HttpResponse, Scope};
+use chrono::Utc;
+
+use diesel::{ExpressionMethods, QueryDsl, RunQueryDsl};
+use jsonschema::JSONSchema;
+use serde_json::json;
+use service_utils::service::types::DbConnection;
+use service_utils::{bad_argument, db_error, result as superposition};
+use superposition_types::User;
+
+use crate::api::custom_types::types::CreateTypeRequest;
+
+pub fn endpoints() -> Scope {
+    Scope::new("")
+        .service(list_types)
+        .service(create_or_update_type)
+        .service(delete_type)
+}
+
+#[put("")]
+async fn create_or_update_type(
+    request: Json<CreateTypeRequest>,
+    db_conn: DbConnection,
+    user: User,
+) -> superposition::Result<HttpResponse> {
+    let DbConnection(mut conn) = db_conn;
+    let _ = JSONSchema::compile(&request.type_schema).map_err(|err| {
+        log::error!(
+            "Invalid jsonschema sent in the request, schema: {:?} error: {}",
+            request.type_schema,
+            err
+        );
+        bad_argument!(
+            "Invalid jsonschema sent in the request, validation error is: {}",
+            err.to_string()
+        )
+    })?;
+    let timestamp = Utc::now().naive_utc();
+    diesel::insert_into(jsonschema_types::table)
+        .values((
+            jsonschema_types::type_schema.eq(request.type_schema.clone()),
+            jsonschema_types::type_name.eq(request.type_name.clone()),
+            jsonschema_types::created_by.eq(user.email.clone()),
+            jsonschema_types::created_at.eq(timestamp),
+        ))
+        .on_conflict(jsonschema_types::type_name)
+        .do_update()
+        .set((
+            jsonschema_types::type_schema.eq(request.type_schema.clone()),
+            jsonschema_types::created_by.eq(user.email),
+        ))
+        .execute(&mut conn)
+        .map_err(|err| {
+            log::error!("failed to insert custom type with error: {}", err);
+            db_error!(err)
+        })?;
+    Ok(HttpResponse::Ok().json(json!({"message": "custom type created successfully"})))
```

Comment:
We should move to sending back the resource on successful creation instead of sending a message back.